pkgs/by-name/in/install-grub-ng/src/builder/entries.rs

```diff
@@ -0,0 +1,424 @@
+			links
+				.map(|d| d.map(|p| p.path()))
+				.collect::<Result<Vec<_>, _>>()?
+		} else {
+			vec![]
+		};
+
+		links.sort();
+
+		if !current && !links.is_empty() {
+			writeln!(
+				&mut self.inner,
+				r#"submenu "> {name}{name_suffix}" --class submenu {{"#
+			)?;
+		}
+
+		self.add_entry(
+			&format!(
+				"{name}{}{name_suffix}",
+				if !links.is_empty() { " - Default" } else { "" }
+			),
+			path,
+			options,
+			current,
+		)?;
+
+		// Find all the children of the current default configuration
+		// Do not search for grand children
+		for link in &links {
+			// FIXME:
+			// This currently returns 1970-01-01 no matter what.
+			// These links seem to be have the correct creation date set, but not their
+			// modified date...?
+			let date = Self::generation_date_from_link(link)?;
+
+			let version = if let Ok(version) = fs::read_to_string(link.join("nixos-version")) {
+				version
+			} else {
+				let modules = link
+					.join("kernel")
+					.canonicalize()?
+					.parent()
+					.ok_or_else(|| {
+						eyre!("Somehow {}/kernel doesn't have a parent..?", link.display())
+					})?
+					.join("lib/modules");
+
+				let Some(version) = fs::read_dir(&modules)?.find_map(|m| {
+					m.ok()
+						.and_then(|p| Some(p.path().file_name()?.to_string_lossy().into_owned()))
+				}) else {
+					bail!("Could not deduce the current NixOS version")
+				};
+
+				version
+			};
+
+			let entry_name =
+				fs::read_to_string(link.join("configuration-name")).unwrap_or_else(|_| {
+					format!(
+						"({} - {date} - {version})",
+						link.file_name().unwrap_or_default().to_string_lossy()
+					)
+				});
+
+			self.add_entry(&format!("{name} - {entry_name}"), link, "", true)?;
+		}
+
+		if !current && !links.is_empty() {
+			writeln!(&mut self.inner, "}}")?;
+		}
+
+		Ok(())
+	}
+
+	fn generation_date_from_link(link: &Path) -> Result<time::Date> {
+		let sys_time = link.symlink_metadata()?.modified()?;
+
+		Ok(time::OffsetDateTime::from(sys_time).date())
+	}
+
+	fn add_entry(&mut self, name: &str, path: &Path, options: &str, current: bool) -> Result<()> {
+		let kernel_dir = path.join("kernel");
+		let initrd_dir = path.join("initrd");
+
+		if !(kernel_dir.exists() && initrd_dir.exists()) {
+			return Ok(());
+		}
+
+		let kernel_dir = self.copy_to_kernels_dir(&kernel_dir)?;
+		let initrd_dir = self.copy_to_kernels_dir(&initrd_dir)?;
+
+		// Include second initrd with secrets
+		let secrets_dir = self
+			.append_initrd_secrets(name, path, current)?
+			.unwrap_or_default();
+
+		// FIXME: $confName
+
+		let kernel_params = format!(
+			"init={} {}",
+			path.join("init").canonicalize()?.display(),
+			fs::read_to_string(path.join("kernel-params"))?
+		);
+
+		let xen = path.join("xen.gz");
+		let xen = if xen.exists() {
+			Some((
+				self.copy_to_kernels_dir(&xen)?,
+				fs::read_to_string(path.join("xen-params")).unwrap_or_default(),
+			))
+		} else {
+			None
+		};
```

Comment:
Has the Xen stuff been tested against #324911? The Perl implementation was broken so we simply dropped BIOS support there.

Reply:
To be honest, I have no idea as I've never used Xen myself. It's probably a good idea to drop this as it is after all a direct port of a purportedly broken script

Reply:
Agreed. The team would rather just support `systemd-boot` and Lanzaboote, so this can be dropped.